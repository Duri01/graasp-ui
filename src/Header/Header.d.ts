import React from 'react'

interface HeaderProps {
  classes: {
    appBar: string
    appBarShift: string
    menuButton: string
    hide: string
  }
  theme: {
    direction: string
  }
  isSidebarOpen: boolean
  handleDrawerOpen: () => {}
}

declare const Header: React.FC<HeaderProps>
