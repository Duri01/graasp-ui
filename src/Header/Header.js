import React from 'react'
import { withStyles } from '@material-ui/core/styles'
import clsx from 'clsx'
import AppBar from '@material-ui/core/AppBar'
import Toolbar from '@material-ui/core/Toolbar'
import IconButton from '@material-ui/core/IconButton'
import MenuIcon from '@material-ui/icons/Menu'
import { DRAWER_WIDTH } from '../constants'

const styles = (theme) => ({
  appBar: {
    transition: theme.transitions.create(['margin', 'width'], {
      easing: theme.transitions.easing.sharp,
      duration: theme.transitions.duration.leavingScreen
    })
  },
  appBarShift: {
    width: `calc(100% - ${DRAWER_WIDTH}px)`,
    marginLeft: DRAWER_WIDTH,
    transition: theme.transitions.create(['margin', 'width'], {
      easing: theme.transitions.easing.easeOut,
      duration: theme.transitions.duration.enteringScreen
    })
  },
  menuButton: {
    marginLeft: theme.spacing(2),
    marginRight: theme.spacing(3)
  },
  hide: {
    display: 'none'
  }
})

const Header = ({ id, classes, handleDrawerOpen, isSidebarOpen }) => {
  return (
    <AppBar
      position='fixed'
      className={clsx(classes.appBar, {
        [classes.appBarShift]: isSidebarOpen
      })}
    >
      <Toolbar disableGutters={!isSidebarOpen}>
        <IconButton
          id={id}
          color='inherit'
          aria-label='Open drawer'
          onClick={handleDrawerOpen}
          className={clsx(classes.menuButton, isSidebarOpen && classes.hide)}
        >
          <MenuIcon />
        </IconButton>
      </Toolbar>
    </AppBar>
  )
}

const StyledComponent = withStyles(styles, { withTheme: true })(Header)

export default StyledComponent
