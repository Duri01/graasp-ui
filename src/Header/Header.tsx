import React, { FC } from 'react';
import { makeStyles } from '@mui/styles';
import clsx from 'clsx';
import AppBar from '@mui/material/AppBar';
import Toolbar from '@mui/material/Toolbar';
import IconButton from '@mui/material/IconButton';
import MenuIcon from '@mui/icons-material/Menu';
import { DRAWER_WIDTH } from '../constants';
import { OPEN_DRAWER_LABEL } from '../texts';

export type HeaderProps = {
  id?: string;
  isSidebarOpen?: boolean;
  handleDrawerOpen?: () => void;
  hasSidebar: boolean;
  openDrawerAriaLabel?: string;
  leftContent?: React.ReactElement;
  rightContent?: React.ReactElement;
};

const useStyles = makeStyles((theme) => ({
  appBar: {
    transition: theme.transitions.create(['margin', 'width'], {
      easing: theme.transitions.easing.sharp,
      duration: theme.transitions.duration.leavingScreen,
    }),
  },
  appBarShift: {
    width: `calc(100% - ${DRAWER_WIDTH}px)`,
    marginLeft: DRAWER_WIDTH,
    transition: theme.transitions.create(['margin', 'width'], {
      easing: theme.transitions.easing.easeOut,
      duration: theme.transitions.duration.enteringScreen,
    }),
  },
  menuButton: {
    marginLeft: theme.spacing(2),
    marginRight: theme.spacing(3),
  },
  hide: {
    display: 'none',
  },
  toolbar: {
    justifyContent: 'space-between',
  },
}));

export const Header: FC<HeaderProps> = ({
  id,
  hasSidebar,
  handleDrawerOpen,
  isSidebarOpen = false,
  openDrawerAriaLabel = OPEN_DRAWER_LABEL,
  leftContent,
  rightContent,
}) => {
  const classes = useStyles();
  return (
    <AppBar
      position='fixed'
      className={clsx(classes.appBar, {
        [classes.appBarShift]: isSidebarOpen,
      })}
    >
      <Toolbar disableGutters={!isSidebarOpen} className={classes.toolbar}>
        {hasSidebar && (
          <IconButton
            id={id}
            color='inherit'
            aria-label={openDrawerAriaLabel}
            onClick={handleDrawerOpen}
            className={clsx(classes.menuButton, {
              [classes.hide]: isSidebarOpen,
            })}
          >
            <MenuIcon />
          </IconButton>
        )}
        {leftContent}
        {rightContent}
      </Toolbar>
    </AppBar>
  );
};

export default Header;
