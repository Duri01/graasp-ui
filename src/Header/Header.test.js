import React from 'react'
import AppBar from '@material-ui/core/AppBar'
import IconButton from '@material-ui/core/IconButton'
import MenuIcon from '@material-ui/icons/Menu'
import { shallow } from 'enzyme'
import { Header } from './Header'

const createHeaderProps = ({ id, isSidebarOpen, hasSidebar } = {}) => {
  return {
    id,
    classes: {
      appBar: '',
      appBarShift: '',
      menuButton: '',
      hide: 'hide'
    },
    hasSidebar,
    handleDrawerOpen: jest.fn(),
    isSidebarOpen
  }
}

describe('Header', () => {
  let wrapper

  describe('default props', () => {
    beforeAll(() => {
      const props = createHeaderProps()
      // eslint-disable-next-line react/jsx-props-no-spreading
      wrapper = shallow(<Header {...props} />)
    })

    it('renders correctly', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('renders one <AppBar /> component without drawer icon', () => {
      const appBar = wrapper.find(AppBar)
      expect(appBar).toHaveLength(1)
      const icon = wrapper.find(IconButton)
      expect(icon).toHaveLength(0)
    })
  })

  describe('hasSidebar = true', () => {
    describe('isSidebarOpen = false', () => {
      beforeAll(() => {
        const props = createHeaderProps({
          hasSidebar: true,
          isSidebarOpen: false
        })
        // eslint-disable-next-line react/jsx-props-no-spreading
        wrapper = shallow(<Header {...props} />)
      })

      it('renders correctly', () => {
        expect(wrapper).toMatchSnapshot()
      })

      it('renders one <AppBar /> component with drawer icon not hidden', () => {
        const appBar = wrapper.find(AppBar)
        expect(appBar).toHaveLength(1)
        const icon = wrapper.find(IconButton)
        expect(icon).toHaveLength(1)
        expect(icon.hasClass('hide')).toEqual(false)
        const menuIcon = icon.find(MenuIcon)
        expect(menuIcon).toHaveLength(1)
      })
    })

    describe('isSidebarOpen = true', () => {
      beforeAll(() => {
        const props = createHeaderProps({
          hasSidebar: true,
          isSidebarOpen: true,
          id: 'id'
        })
        // eslint-disable-next-line react/jsx-props-no-spreading
        wrapper = shallow(<Header {...props} />)
      })

      it('renders one <AppBar /> component with drawer icon hidden', () => {
        const appBar = wrapper.find(AppBar)
        expect(appBar).toHaveLength(1)
        const icon = wrapper.find(IconButton)
        expect(icon).toHaveLength(1)
        expect(icon.prop('id')).toEqual('id')
        expect(icon.hasClass('hide')).toEqual(true)
        const menuIcon = icon.find(MenuIcon)
        expect(menuIcon).toHaveLength(1)
      })
    })
  })
})
