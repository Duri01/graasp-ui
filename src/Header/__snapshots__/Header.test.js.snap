// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header default props renders correctly 1`] = `
<WithStyles(ForwardRef(AppBar))
  className=""
  position="fixed"
>
  <WithStyles(ForwardRef(Toolbar))
    disableGutters={true}
  />
</WithStyles(ForwardRef(AppBar))>
`;

exports[`Header hasSidebar = true isSidebarOpen = false renders correctly 1`] = `
<WithStyles(ForwardRef(AppBar))
  className=""
  position="fixed"
>
  <WithStyles(ForwardRef(Toolbar))
    disableGutters={true}
  >
    <WithStyles(ForwardRef(IconButton))
      aria-label="sometext"
      className=""
      color="inherit"
      onClick={[MockFunction]}
    >
      <MenuIcon />
    </WithStyles(ForwardRef(IconButton))>
  </WithStyles(ForwardRef(Toolbar))>
</WithStyles(ForwardRef(AppBar))>
`;
